src/opcua-connector/machineState.ts

```diff
@@ -0,0 +1,132 @@
+import deepEqual from 'deep-equal'
+import { combineLatest } from 'rxjs'
+import { map } from 'rxjs/internal/operators/map'
+import { debounceTime, distinctUntilChanged, tap } from 'rxjs/operators'
+import { OpcuaStreams, Rule, Rules, VariableStream, VariableStreamData } from './types'
+import * as uuid from 'uuid'
+
+type MachineStateEvent = {
+  state: number
+  stateDesc?: string
+  generateError: boolean
+}
+type MachineStateEvents = Array<MachineStateEvent>
+
+type MachineStateEventSource = {
+  values: VariableStreamData
+  name: string
+  rule: Rule
+}
+
+type EmitStateHandler = (state: number, description: string | undefined) => void
+type EmitErrorHandler = (id: string, error: number, description: string | undefined) => void
+
+/**
+ * Listen to all values in the stream and apply the rules if one of them changed
+ * There is a little debounce to get all values who are in the same query interval
+ *
+ * @param streams stream of all opcua values
+ * @param rules rules from the settings how an when to emit events
+ * @param emitState callback when a state event needs to be emitted
+ * @param emitError callback when an error event needs to be emitted
+ */
+export const executeStateEmitter = (
+  streams: OpcuaStreams,
+  rules: Rules,
+  emitState: EmitStateHandler,
+  emitError: EmitErrorHandler,
+): void => {
+  combineLatest(streams)
+    .pipe(
+      debounceTime(10),
+      map<VariableStream, VariableStreamData>(toValues),
+      map((values) => Object.entries(rules).map(validateRule(values)).filter(notUndefined)),
+      map((events) => events.map(renderDescription)),
+      distinctUntilChanged(deepEqual),
+      tap(publishEvents(emitState, emitError)),
+    )
+    .subscribe()
+}
+
+/** convert the variants to the real values as any type */
+const toValues = (values: VariableStream): VariableStreamData =>
+  Object.entries(values).reduce(
+    (acc, [key, v]) => ({ ...acc, [key]: v.value }),
+    {},
+  ) as VariableStreamData
+
+/**
+ * Validate the given rule with the value stream
+ *
+ * It will create a function, executing the code giving in the settings
+ *
+ * WARNING: this code is not validating the settings for any vulnerability code.
+ *
+ */
+const validateRule =
+  (values: VariableStreamData) =>
+  ([name, rule]: [string, Rule]): MachineStateEventSource | undefined => {
+    const vars = Object.entries(values).reduce(
```

Comment:
It's good you included the `malicious code` warning. 
Perhaps we can find a simple DSL lib that sanitizes input?

Reply:
I'm very afraid that we are going to overengineer something here.